trunk/SUAVE/Methods/Aerodynamics/Airfoil_Panel_Method/airfoil_analysis.py

```diff
@@ -0,0 +1,573 @@
+    first_panel     = list(prev_index.flatten())
+    last_panel      = list((first_idx-1 + mask_count).flatten())
+    last_paneldve   = list((first_idx-2 + mask_count).flatten())
+    aoas            = list(np.repeat(np.arange(nalpha),nRe))
+    res             = list(np.tile(np.arange(nRe),nalpha) )
+    X_BOT.mask[first_panel,aoas,res] = False
+    
+    # flow velocity and pressure of on botton surface 
+    VE_BOT          = -VT[::-1] 
+    CP_BOT          = 1 - VE_BOT**2  
+    
+    # velocity gradients on bottom surface  
+    DVE_BOT                        = np.ma.zeros(np.shape(X_BOT)) 
+    DVE_BOT_TEMP                   = np.diff(VE_BOT,axis = 0)/np.diff(X_BOT,axis = 0)
+    a                              = X_BOT[1:-1] - X_BOT[:-2]
+    b                              = X_BOT[2:]   - X_BOT[:-2]
+    DVE_BOT[1:-1]                  = ((b*DVE_BOT_TEMP[:-1] + a*DVE_BOT_TEMP[1:])/(a+b)).data
+    DVE_BOT.mask                   = X_BOT.mask
+    DVE_BOT[first_panel,aoas,res]  = DVE_BOT_TEMP[first_panel,aoas,res]
+    DVE_BOT[last_panel,aoas,res]   = DVE_BOT_TEMP[last_paneldve,aoas,res] 
+    
+    # x - location of stagnation point 
+    L_BOT                          = X_BOT[-1,:,:]    
+        
+    # laminar boundary layer properties using thwaites method 
+    BOT_T_RESULTS  = thwaites_method(nalpha,nRe, L_BOT , RE_L_VALS, X_BOT, VE_BOT, DVE_BOT,batch_analysis,
+                                     THETA_0=1E-5,n=n_computation) 
+    X_T_BOT          = BOT_T_RESULTS.X_T      
+    THETA_T_BOT      = BOT_T_RESULTS.THETA_T     
+    DELTA_STAR_T_BOT = BOT_T_RESULTS.DELTA_STAR_T  
+    H_T_BOT          = BOT_T_RESULTS.H_T         
+    CF_T_BOT         = BOT_T_RESULTS.CF_T   
+    RE_THETA_T_BOT   = BOT_T_RESULTS.RE_THETA_T    
+    RE_X_T_BOT       = BOT_T_RESULTS.RE_X_T      
+    DELTA_T_BOT      = BOT_T_RESULTS.DELTA_T      
+     
+    # transition location  
+    TR_CRIT_BOT     = RE_THETA_T_BOT - 1.174*(1 + 224000/RE_X_T_BOT)*RE_X_T_BOT**0.46  
+    CRITERION_BOT   = np.ma.masked_greater(TR_CRIT_BOT,0 ) 
+    if batch_analysis:
+        mask_count  = np.ma.count(CRITERION_BOT,axis = 0)  
+        mask_count[mask_count == n_computation] = n_computation-1
+    else:
+        mask_count  =  np.zeros((nalpha,nalpha)).astype(int) 
+    transition_panel = list(mask_count.flatten()) 
+    aoas             = list(np.repeat(np.arange(nalpha),nRe))
+    res              = list(np.tile(np.arange(nRe),nalpha))
+        
+    X_TR_BOT          = X_T_BOT[transition_panel,aoas,res].reshape(nalpha,nRe)
+    DELTA_STAR_TR_BOT = DELTA_STAR_T_BOT[transition_panel,aoas,res].reshape(nalpha,nRe)
+    THETA_TR_BOT      = THETA_T_BOT[transition_panel,aoas,res].reshape(nalpha,nRe)    
+    DELTA_TR_BOT      = DELTA_T_BOT[transition_panel,aoas,res].reshape(nalpha,nRe) 
+    
+    TURBULENT_SURF  = L_BOT.data  - X_TR_BOT
+    TURBULENT_COORD = X_BOT       - X_TR_BOT
+    
+    # turbulent boundary layer properties using heads method 
```

Comment:
Because this is a 3d calculation i.e. each airfoil at each angle of attack and each Reynolds number, the fasting to do is a mask than a nested for loop.